src/main/kotlin/org/taktik/icure/utils/entities/embed/TimeTableItem.kt

```diff
@@ -104,26 +105,15 @@ fun TimeTableItem.iterator(startDateAndTime: Long, endDateAndTime: Long, duratio
 		}
 }
 
-fun List<TimeTableHour>.iterator(duration: Duration) = object : Iterator<Long> {
-	var current: LocalTime? = this@iterator.firstOrNull()?.let { (it.startHour ?: 0L).toLocalTime() } ?: throw NoSuchElementException()
-	var idx = 0
-
-	init {
-		if (!hasNext()) throw NoSuchElementException()
-	}
+fun List<TimeTableHour>.iterator(duration: Duration): Iterator<Long> = this.map { it.iterator(duration) }.sortedMerge()
+fun TimeTableHour.iterator(duration: Duration) = object : Iterator<Long> {
+	var current: LocalTime? = (this@iterator.startHour ?: 0L).takeIf { (this@iterator.endHour ?: 0L) == 0L || it < (this@iterator.endHour ?: 240000) }?.toLocalTime()
 
 	override fun hasNext() = current != null
 	override fun next() = (current ?: throw NoSuchElementException()).let { c ->
 		c.toLongHour().also {
-			current = (c + duration).let {
-				when {
-					it.toLongHour() < (this@iterator[idx].endHour ?: 240000) -> it
-					idx < this@iterator.size - 1 -> {
-						idx++
-						(this@iterator[idx].startHour ?: 0L).toLocalTime()
-					}
-					else -> null
-				}
+			current = (c + duration).takeIf { it > c }?.let {
```

Comment:
pretty sure you can make a single function for the update and initialisation of current